### DISCUSSION TITLE
Setup portable code multiplatform ImPlatform.h

### DISCUSSION THREAD

**Discussion by soufianekhiat** [ocornut/imgui#3921] [Old/Archived discussions]
Open question:
How could we setup a portable example with minimum dependencies?

Currently, when we want to share code using ImGui we have to make choose of one backend or reimplement (copy paste examples) for each target.

Does that make sense to have a way to move platform-independent code out of our main?

The re-routing will be done from ImPlatform.h, which mostly include backend under #ifdefs.
Which can fit perfectly on a build system (Sharpmake, CMake, ...), we can dispatch various builds with specification of global defines "IS_WINDOWS", "IS_APPLE", ...

ImPlatform could be splitted:
- ImWindow.h + ImWindow.cpp
- ImGfxAPI.h + ImGfxAPI.cpp

A draft of a target code:
```
#ifdef IS_WINDOWS
#define IMGUI_TARGET_WIN32
#define IMGUI_TARGET_OPENGL3
#elif IS_APPLE
#define IMGUI_TARGET_OSX
#define IMGUI_TARGET_METAL
#elif IS_APPLE
#define IMGUI_TARGET_GLFW
#define IMGUI_TARGET_VULKAN
#endif
#include <ImPlatform.h>

int main()
{
    ImInitWindow();
    ImInitGfxAPI();

    // LoadFont...
    while (ImPlatformContinue())
    {
        ImPlatformEvens();

        ImGfxAPINewFrame();
        ImPlatformNewFrame();

        // ImGui User Code

        ImGui::Render();

        ImGfxAPIClear();
        // User GfxCode

        ImGfxAPIRender(ImGui::GetDrawData());

        ImGfxAPISwapBuffer();
    }

    ImShutDownGfxAPI();
    ImShutDownWindow();

    return 0;
}

```

**Comment by ocornut** [comment]
We have such helper internally, but releasing it would create much more support request for us so it's presently not worth it.

**Reply by soufianekhiat** [reply]
It's totally understandable.
Did you consider having an "unsupported" folder. That is a strategy used by Eigen (Linear Algebra Library), everything on the subfolder only received support from community "as-is" otherwise no official support.

**Comment by soufianekhiat** [comment]
Here my attempt for that (For now only support {Win32}x{Dx9, Dx10, Dx11, OpenGL3}:
https://github.com/soufianekhiat/DearWidgets/tree/ImPlatform
https://github.com/soufianekhiat/DearWidgets/blob/ImPlatform/src/api/ImPlatform.h
https://github.com/soufianekhiat/DearWidgets/blob/ImPlatform/src/api/ImPlatform.cpp
// Mostly copy past from imgui/examples

The main is simplified as:
https://github.com/soufianekhiat/DearWidgets/blob/ImPlatform/src/demo/demo.cpp#L118

TODO: Add more platform {Linux, Emscriptem, Mac} more Gfx Api {Vulkan, Metal, WGPU}.
Support for more ImGui feature like: Viewport {Not functional on Win32-OpenGL3}
